### ISSUE TITLE
Confusing Accuracy difference in DGL

### ISSUE THREAD

**Issue by Sinasta** [wassimj/topologicpy#10]
When i train with GCN_Classic model on my dataset i get an accuracy of 40%.
But when i replace the GCN_Classic model:

https://github.com/wassimj/topologicpy/blob/08a417dafa065e7963889758818741e7339d59b6/DGL.py#L141-L176

with this:

```
class GCN_Classic(nn.Module):
    def __init__(self, in_feats, h_feats, num_classes):
        super(GCN_Classic, self).__init__()
        self.conv1 = GraphConv(in_feats, h_feats[0])
        self.conv2 = GraphConv(h_feats[0], num_classes)

    def forward(self, g, in_feat):
        h = self.conv1(g, in_feat)
        h = F.relu(h)
        h = self.conv2(g, h)
        g.ndata['h'] = h
        return dgl.mean_nodes(g, 'h')
```

I get an accuracy of 80%
I must admit that i am extremely confused by this difference in accuracy because both models should technically yield the exact same results when called with hl_widths=[32] (one single hidden layer). As far as i understand the only difference in your implementation is the possibility to add multiple hidden layers.
I will investigate a bit further but i wanted to open this issue to ask you if you have an explanation for this difference in accuracy.
P.s. just to clarify: i did not change anything other then the lines mentioned

**Comment by wassimj** [comment]
Hmm.. I am not seeing what the difference is for one hidden layer. Do you get bad accuracy if you use SAGEConv?

**Comment by wassimj** [comment]
Can you share all your hyperparameters and datasets? Are you using the dataset that I provide or your own?

**Comment by wassimj** [comment]
I just did some testing and implemented your code and compared. In my case with exactly the same dataset, hyperparameters and just changing the cv_type from "classic" to "raban" I get:
classic (mine): 58% accuracy on unseen data
raban (yours): 37% accuracy on unseen data

So I am not able to replicate your results. Please do more testing or send me all the information so I can test on my side. Thanks

**Comment by wassimj** [comment]
Here are my hyper-parameters:
optimizer = DGL.Optimizer()
hparams = DGL.Hyperparameters(optimizer=optimizer,
                              cv_type="Holdout",
                              k_folds=2,
                              split=0.2,
                              hl_widths=[32],
                              conv_layer_type="raban",
                              epochs=30,
                              pooling="MaxPooling",
                              loss_function="Cross Entropy",
                              batch_size=100,
                              classifier_path=r"C:\Users\wassimj\dgl_classifier.pt",
                              results_path=r"C:\Users\wassimj\dgl_results.csv")

**Comment by wassimj** [comment]
I pushed the code that I am using for testing. You can now inspect it and use either "classic" or "raban". Please report back. Thank you for your help in debugging the code.

**Comment by Sinasta** [comment]
> I just did some testing and implemented your code and compared. In my case with exactly the same dataset, hyperparameters and just changing the cv_type from "classic" to "raban" I get: classic (mine): 58% accuracy on unseen data raban (yours): 37% accuracy on unseen data
> 
> So I am not able to replicate your results. Please do more testing or send me all the information so I can test on my side. Thanks

Thank you for your answer. I am not able to investigate further at the moment but don't you think that even your results show that something is wrong? Technically (if i am not mistaken) your two tests should yield the exact same accuracy. It should not matter if you use cv_type = "classic" or cv_type = "raban" because the model should be the exact same. That is what confuses me.
I hope to give you more information tomorrow.

**Comment by wassimj** [comment]
The network is initialised with random weights and I don’t set the same seed so the performance, before training proceeds too much, will vary. Both should settle on similar performance at the end. It might be a fluke that you got 80% on first shot but I don’t know how many epochs and all the rest of hyperparams that you used.

**Comment by Sinasta** [comment]
I did some more debugging and i think i found the problem and the fix:
as shown [here](https://discuss.dgl.ai/t/number-of-layers-in-dgl-example-gat/2734) and [here](https://discuss.dgl.ai/t/how-many-layers-do-you-use-for-gcn-based-on-dgl/124) when adding hidden layers to a list, the list needs to be initialized as [`nn.ModuleList()`](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html)

so when i change those lines:
https://github.com/wassimj/topologicpy/blob/0c5b25c621528b993647fc4343e5394129a5cb25/DGL.py#L163
https://github.com/wassimj/topologicpy/blob/0c5b25c621528b993647fc4343e5394129a5cb25/DGL.py#L197
https://github.com/wassimj/topologicpy/blob/0c5b25c621528b993647fc4343e5394129a5cb25/DGL.py#L238
https://github.com/wassimj/topologicpy/blob/0c5b25c621528b993647fc4343e5394129a5cb25/DGL.py#L279
https://github.com/wassimj/topologicpy/blob/0c5b25c621528b993647fc4343e5394129a5cb25/DGL.py#L319
to: `self.list_of_layers = nn.ModuleList()`

everything seems to work as expected and i get above 80 percent accuracy with every test.
Please let me know what you think of that!

**Comment by wassimj** [comment]
Excellent! Thank you very much for finding this! I never realised that nn.ModuleList() is needed and not just a regular python list. I have pushed to github a new version. Please test and let me know so I can close this and issue a new release of topologicpy. Thanks again for all your help.

**Comment by Sinasta** [comment]
I am always amazed by how responsive and fast you are in fixing bugs and pushing commits. I am glad that i could be of help. I am testing the new code currently and to me it seems that everything is working fine now. This issue can be closed and i will reopen as soon as something is not working correctly. Thanks a lot!
